NeoML/src/Dnn/DnnDistributed.cpp

```diff
@@ -508,47 +508,72 @@ void CDistributedTraining::StoreDnn( CArchive& archive, int index, bool storeSol
 
 //---------------------------------------------------------------------------------------------------------------------
 
-void CDistributedInference::initialize( int threadsCount )
+// Params to transfer to all threads function
+struct CDistributedInference::CThreadParams final {
+	IDistributedDataset* Data = nullptr; // Pointer to data for the inference for all dnns
+	CObjectArray<CDnnReference> Refs; // Separate dnn for each thread
+	CArray<bool> IsDnnInferenced; // Indicates for what dnns the inference was performed
+	CArray<CString> ErrorMessages; // Containers for errors if it happened
+	bool IsErrorHappened = false;
+
+	CThreadParams( int threadsCount, CReferenceDnnFactory& referenceDnnFactory );
+	void Initialize( IDistributedDataset& data );
+};
+
+CDistributedInference::CThreadParams::CThreadParams( int threadsCount, CReferenceDnnFactory& referenceDnnFactory )
 {
 	initThreadGroupInfo();
 
-	threadParams.ErrorMessages.Add( {}, threadsCount );
 	// Create reference dnns
 	// To create a reference dnn the original network should be trained or at least reshaped
 	// All training paramBlobs should exist
-	threadParams.Refs.SetBufferSize( threadsCount );
+	Refs.SetBufferSize( threadsCount );
 	for( int i = 1; i < threadsCount; ++i ) {
-		threadParams.Refs.Add( referenceDnnFactory->CreateReferenceDnn() );
+		Refs.Add( referenceDnnFactory.CreateReferenceDnn() );
 	}
 	// Here it can be either a one more reference dnn
 	// Or also the original dnn, because no one can create a new reference dnn, while this inference
-	threadParams.Refs.Add( referenceDnnFactory->CreateReferenceDnn( /*getOriginalDnn*/true ) );
+	Refs.Add( referenceDnnFactory.CreateReferenceDnn( /*getOriginalDnn*/true ) );
 }
 
+void CDistributedInference::CThreadParams::Initialize( IDistributedDataset& data )
+{
+	Data = &data;
+	IsDnnInferenced.DeleteAll();
+	IsDnnInferenced.Add( false, Refs.Size() );
+	ErrorMessages.SetSize( Refs.Size() );
```

Comment:
This line doesn't cleanup old error messages

Reply:
fixed
I wanted avoid iterate through `ErrorMessages` to speed-up, but I understand now, that we do not want to see old  `ErrorMessages` for new cases.